processing/src/main/java/org/apache/druid/query/groupby/ResultRowObjectMapperDecoratorUtil.java

```diff
@@ -0,0 +1,236 @@
+            for (int i = 0; i < queryDimensions.size(); ++i) {
+              if (dimensionRequiresConversion(queryDimensions.get(i).getOutputType())) {
+                final int dimensionIndexInResultRow = query.getResultRowDimensionStart() + i;
+                resultRow.set(
+                    dimensionIndexInResultRow,
+                    objectMapper.convertValue(
+                        resultRow.get(dimensionIndexInResultRow),
+                        javaTypes[dimensionIndexInResultRow]
+                    )
+                );
+              }
+            }
+
+            return resultRow;
+          } else {
+            if (!jp.isExpectedStartArrayToken()) {
+              throw DruidException.defensive("Expected start token, received [%s]", jp.currentToken());
+            }
+
+            Object[] objectArray = new Object[query.getResultRowSizeWithPostAggregators()];
+            int index = 0;
+
+            while (jp.nextToken() != JsonToken.END_ARRAY) {
+              objectArray[index] = JacksonUtils.readObjectUsingDeserializationContext(jp, ctxt, javaTypes[index]);
+              ++index;
+            }
+
+            return ResultRow.of(objectArray);
+          }
+        }
+      };
+    }
+  }
+
+  @Nullable
+  private static JsonSerializer<ResultRow> getSerializer(
+      final GroupByQuery query,
+      final GroupByQueryConfig groupByQueryConfig
+  )
+  {
+    final boolean resultAsArray = query.context().getBoolean(GroupByQueryConfig.CTX_KEY_ARRAY_RESULT_ROWS, false);
+    final boolean intermediateCompatMode = groupByQueryConfig.isIntermediateResultAsMapCompat();
+    final boolean arrayBasedRows = resultAsArray && !intermediateCompatMode;
+    if (arrayBasedRows) {
+      return null;
+    } else {
+      if (resultAsArray) {
+        return new JsonSerializer<ResultRow>()
+        {
+          @Override
+          public void serialize(ResultRow resultRow, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
+              throws IOException
+          {
+            JacksonUtils.writeObjectUsingSerializerProvider(jsonGenerator, serializerProvider, resultRow.getArray());
+          }
+        };
+
+      } else {
+        return new JsonSerializer<ResultRow>()
+        {
+          @Override
+          public void serialize(ResultRow resultRow, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
+              throws IOException
+          {
+            JacksonUtils.writeObjectUsingSerializerProvider(
+                jsonGenerator,
+                serializerProvider,
+                resultRow.toMapBasedRow(query)
+            );
+          }
+        };
+      }
+    }
+  }
+
+  private static boolean dimensionRequiresConversion(final ColumnType dimensionType)
+  {
+    return dimensionType.is(ValueType.COMPLEX) && !ColumnType.NESTED_DATA.equals(dimensionType);
+  }
+
+  private static JavaType[] createJavaTypesForResultRow(final GroupByQuery groupByQuery)
```

Comment:
ah i see, inside the version that takes a class its going to call typeFactory.constructType anyway, :+1: